File: 2017/Qualifications_A/NodeJS/PancakeRunner.js
class PancakeRunner {
    constructor() {
        this.happy = '+';
        this.notHappy = '-'; 
    }

    process(run) {
        let flipperSize = run.flipperSize;
        let data = [...run.data];

        let count = 0;
        let size = data.length - flipperSize + 1;
        for(let loop = 0; loop < size; loop++) {
            if(data[loop] === this.notHappy) {
                this.flip(data, flipperSize, loop);
                count++;
            }
        }
            
        return {index: run.index, success: this.isAllHappy(data), result: count };
    }

    flip(data, flipperSize, startIndex) {
        for(let loop = startIndex; loop < startIndex + flipperSize; loop++) {
            data[loop] = data[loop] === this.notHappy ?
                this.happy : this.notHappy;
        }
    }

    isAllHappy(data) {
        return data.every(x => x === this.happy);
    }
}

module.exports = PancakeRunner;

File: 2017/Qualifications_A/DotNet/PancakeRunner.cs
using System.Collections;
using System.Linq;
using System.Text;

namespace CodeJam
{
    public class PancakeRunner
    {
        private const char Happy = '+';
        private const char NotHappy = '-';

        public PancakeResult Process(PancakeRun run)
        {
            var res = Process(run.Data, run.FlipperSize);
            return new PancakeResult(run, res);
        }

        private int? Process(string state, int flipperSize)
        {
            if (IsSuccess(state))
            {
                return 0;
            }

            var currentState = state;
            var flipCount = 0;

            for (int loop = 0; loop < state.Length; loop++)
            {
                if (currentState[loop] == NotHappy)
                {
                    currentState = TryFlip(currentState, loop, flipperSize);
                    flipCount++;

                    if (currentState == null)
                    {
                        return null;
                    }

                    if (IsSuccess(currentState))
                    {
                        return flipCount;
                    }
                }
            }

            return null;
        }

        private static string TryFlip(string state, int start, int flipperSize)
        {
            if (start + flipperSize > state.Length)
            {
                return null;
            }

            var newState = new StringBuilder(state);
            for (int loop = start; loop < start + flipperSize; loop++)
            {
                newState[loop] = newState[loop] == NotHappy ? Happy : NotHappy;
            }
            return newState.ToString();
        }

        private static bool IsSuccess(string state)
        {
            return state.All(x => x != NotHappy);
        }
    }
}
